src/RequestHandler/outofprocess/websockethandler.cxx

```diff
@@ -208,25 +211,33 @@ WEBSOCKET_HANDLER::IndicateCompletionToIIS(
 
 --*/
 {
-    /*DebugPrintf (ASPNETCORE_DEBUG_FLAG_INFO,
-            "WEBSOCKET_HANDLER::IndicateCompletionToIIS");*/
-
-    _pHandler->SetStatus(FORWARDER_DONE);
+    DebugPrintf(ASPNETCORE_DEBUG_FLAG_INFO,
+        "WEBSOCKET_HANDLER::IndicateCompletionToIIS called %d", _dwOutstandingIo);
 
-    // do not call IndicateCompletion here
-    // wait for handle close callback and then call IndicateCompletion
-    // otherwise we may release W3Context too early and cause AV
-    //_pHttpContext->IndicateCompletion(RQ_NOTIFICATION_PENDING);
+    //
     // close Websocket handle. This will triger a WinHttp callback
     // on handle close, then let IIS pipeline continue.
-    WinHttpCloseHandle(_hWebSocketRequest);
+    // Make sure no pending IO as there is no IIS websocket cancelation,
+    // any unexpected callback will lead to AV. Revisit it once CanelOutGoingIO works
+    //
+    if (_hWebSocketRequest != NULL && _dwOutstandingIo == 0)
+    {
+        DebugPrintf(ASPNETCORE_DEBUG_FLAG_INFO,
+            "WEBSOCKET_HANDLER::IndicateCompletionToIIS");
+
+        _pHandler->SetStatus(FORWARDER_DONE);
+        _fHandleClosed = TRUE;
+        WinHttpCloseHandle(_hWebSocketRequest);
+        _hWebSocketRequest = NULL;
+    }
 }
 
 HRESULT
 WEBSOCKET_HANDLER::ProcessRequest(
     FORWARDING_HANDLER *pHandler,
     IHttpContext *pHttpContext,
-    HINTERNET     hRequest
+    HINTERNET     hRequest,
+    BOOL*         pfHandleCreated
```

Comment:
This bool is really representing if we upgraded to websockets or not.

Reply:
A WinHttp websocket handle is created during upgrade